Create .gitattributes file for auto-formatting of files

awemany opened the pull request:
To improve upon the .formatted-files logic. See commit message for details.

ftrader commented on src/Makefile.am:
```diff
@@ -480,7 +480,7 @@ endif
 check-formatting:
 	@echo "Checking source formatting style..."
-	(cd $(top_srcdir)/src; ../contrib/devtools/clang-format.py check clang-format-3.8 `cat ./.formatted-files`)
+	(cd $(top_srcdir)/src; ../contrib/devtools/clang-format.py check clang-format-3.8 `cat ./.gitattributes|cut -d ' ' -f 1`)
```
This seems to be assuming that all files in .gitattributes shall be subjected to bitcoin-clang-format filter.

I would suggest filtering for those files explicitly when constructing the list to pass.

awemany replied:
Excellent point. Will rework. Is a grep in the pipe good enough here?

EDIT: Added grep.

ftrader replied:
Seems alright as a solution.

gandrewstone commented on doc/developer-notes.md:
````diff
@@ -40,11 +40,23 @@ class Class
+  Add the following to your $HOME/.gitconfig or .git/config in the Bitcoin directory:
+  ```
+[filter "bitcoin-clang-format"]
+        clean = "clang-format-3.8 -style=file %f|contrib/devtools/trailing-comment.py 120"
````
trailing-comment.py needs to run before clang-format.  But best of all, call contrib/devtools/clang-format.py.  Using this script will ensure that your code always matches make check-formatting...

awemany replied:
Ok, I have modified the PR to do so. This needed a change to the clang-format.py script, which I have now added as a separate commit.

gandrewstone commented:
I dislike that a nice, self-documenting filename like .formatted-files becomes .gitattributes.  Especially for a cool but very obscure feature that will likely be used by few people.  Is there any way to keep the original name?  Symlink?

ftrader commented:
What about a general rule in the .gitattributes for all .h / .cpp files under src/ , that call a script which performs the action only if the file is listed in .formatted-files ?

Long-term we should be getting to where we process all C++ source files anyway...

awemany commented:
@thezerg: A symlink should be possible, yes. Probably with some explanation at the top of gitattributes (I think #-comments work there) that if you reach there from .formatted-files, the reason why it looks so funny now is explained. I have added that now.

@ftrader: 
> What about a general rule in the .gitattributes for all .h / .cpp files under src/ , that call a script which performs the action only if the file is listed in .formatted-files ?

Hmm, but that appears to me to duplicate the effort of working through lists of files. Why another indirection here?

ftrader commented:
I would be opposed to symlinks. They should be avoided within the project sources - they are not a very portable feature, and to the extent that our build uses any (I did find some use in 3rd party libs), I think it's regrettable.

>Hmm, but that appears to me to duplicate the effort of walking through lists of files. Why another indirection here?

Two reasons:
- to preserve the neat filenaming that thezerg likes
- to avoid using the VCS-specific file for this as much as possible. The less specific info we need in the .gitattributes, the better it would be I feel. Might make it easier to hook the checks into other systems.

awemany committed:
devtools/clang-format.py: Add 'format-stdout' and 'format-stdout-if-wanted'

The first option 'format-stdout' will read the specified file
and write the formatted result to stdout.

The second variant, 'format-stdout-if-wanted' checks the
file src/.formatted-files and only reformats if the file is
listed in there - or else passes the file as it is on stdout.

This is meant to be used as a .gitattributes 'clean' filter.

A couple of minor tweaks/fixes have been made as well.

awemany commented:
>    to avoid using the VCS-specific file for this as much as possible. The less specific info we need in the .gitattributes, the better it would be I feel. Might make it easier to hook the checks into other systems.

Ok that's a fair point. I'll wait for @thezerg for the final decision on this.

gandrewstone commented:
I like ftrader's idea of wildcarding in .gitattributes and having a script in devtools that checks .formatted-files for the file.  That idea has 2 additional advantages: it gives people a "hook" to add other check-in tests (without having to know anything git specific), and lets a person run the check-in tests manually before actually checking in.

gandrewstone commented:
changing to WIP for my own tracking, let me know when you make changes...

awemany committed:
Create .gitattributes file for auto-formatting of files

Augments the .formatted-files list with a .gitattributes list
that allows to auto-format files upon check-in by a git
clean/smudge filter.

To do so, add the following section to your ~/.gitconfig:

[filter "bitcoin-clang-format"]
    clean = "contrib/devtools/clang-format.py format-stdout-if-wanted clang-format-3.8 %f"
    smudge = cat

Also adds a short note on how to do it to the developer docs.

awemany commented:
I have updated the commit. It now checks the list in .formatted-files. As git has to do that repeatedly, a minor downside here is that the commands that deal with auto-formatting become noticeably slower.

awemany commented:
The failing test here is proxy_test.py, which appears to be completely unrelated to my code.

sickpig commented:
@awemany 
rekick travis job.

sickpig commented:
@awemany 
you are good to go now :smile: 
sometimes we have some spurious failure on travis (this seems to be something we share with a lot of other projects using it).

sickpig commented:
@awemany do you think this is still WIP or that are ready to be merged? (fwiw is ready to be merged imho)

awemany commented:
@sickpig: I am still moderately unhappy with the reduced performance now (and have to look into improving that at some point) but this still will make it a lot quicker and easier for me to fulfill the formatting requirements. So yes, I also think it is ready to be merged.

sickpig commented:
approved

sickpig commented:
@gandrewstone I think this ready to merge and a really needed change.

gandrewstone commented:
Great, glad to see this come in! Please kick me on slack if an older PR is taken out of WIP.
